sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.cosmos.spark
+
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.encoders.{ExpressionEncoder, RowEncoder}
+import org.apache.spark.sql.types.StructType
+
+import java.time.Instant
+import java.util.concurrent.{ConcurrentLinkedQueue, Executors, TimeUnit}
+import java.util.concurrent.atomic.AtomicLong
+import scala.collection.concurrent.TrieMap
+
+/**
+ * Spark serializers are not thread-safe - and expensive to create (dynamic code generation)
+ * So we will use this object pool to allow reusing serializers based on the targeted schema.
+ * The main purpose for pooling serializers (vs. creating new ones in each PartitionReader) is for Structured
+ * Streaming scenarios where PartitionReaders for the same schema could be created every couple of 100
+ * milliseconds
+ * A clean-up task is used to purge serializers for schemas which weren't used anymore
+ * For each schema we have an object pool that will use a soft-limit to limit the memory footprint
+ */
+private object RowSerializerPool {
+  val MaxPooledSerializerCount = 256
+  private[this] val cleanUpIntervalInSeconds = 300
+  private[this] val expirationIntervalInSeconds = 1800
+  private[this] val schemaScopedSerializerMap =
+    new TrieMap[StructType, RowSerializerQueue]
+  private[this] val executorService = Executors.newSingleThreadScheduledExecutor(SparkUtils.daemonThreadFactory())
+
+  executorService.scheduleWithFixedDelay(
```

Comment:
do we need to schedule more cleanup if no new row encoder is left?

can we schedule the next cleanup() in getOrCreateSerializer()? basically we can check if there is any cleanup scheduled if not schedule one.

That way if row encoders are not used for a while we don't have schedule new ones.

Reply:
The cleanup only happens every 5 minutes - don't think it is necessary to optimize perf/overhead for it. IMO keeping it as simple as possible is sufficient here.

Reply:
we can do later if needed. LGTM.